scipy/sparse/csgraph/_disjoint_set.py

```diff
@@ -0,0 +1,138 @@
+"""
+Disjoint set data structure
+"""
+
+
+import numpy as np
+
+
+class DisjointSet:
+    """ Disjoint set data structure for incremental connectivity queries.
+
+    .. versionadded: 1.6.0
+
+    Attributes
+    ----------
+    n_nodes : int
+        The number of nodes in the set.
+    n_components : int
+        The number of components/subsets.
+
+    Methods
+    -------
+    union
+    find
+
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    data structure. The *find* method implements the *path compression*
+    variant. The *union* method implements the *union by size* variant.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+    See Also
+    --------
+    connected_components : Analyze the connected components of a sparse graph.
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import DisjointSet
+
+    Initialize a disjoint set:
+
+    >>> dis = DisjointSet()
+
+    Merge some subsets:
+
+    >>> dis.union(0, 1)
+    True
+    >>> dis.union(2, 3)
+    True
+    >>> dis.union(3, 3)
+    False
+
+    Find a root node:
+
+    >>> dis.find(1)
+    0
+
+    """
+    def __init__(self):
+        self.n_nodes = 0
+        self.n_components = 0
+        self._sizes = {}
+        self._parents = {}
+
+    def find(self, x):
+        """Find the root node of `x`.
+
+        Parameters
+        ----------
+        x : int
+            Input node.
+
+        Returns
+        -------
+        root : int
+            Root node of `x`.
+        """
+        if not np.issubdtype(type(x), np.integer):
+            raise TypeError("`x` must be an integer")
+
+        if x not in self._parents:
+            self._sizes[x] = 1
+            self._parents[x] = x
+            self.n_nodes += 1
+            self.n_components += 1
+            return x
+
+        parents = self._parents
+        parent = parents[x]
+        while parent != parents[parent]:
+            parent = parents[parent]
+        parents[x] = parent
+        return parent
+
+    def union(self, a, b):
+        """Merge the subsets of `a` and `b`.
+
+        The smaller subset (the child) is merged into the larger subset (the
+        parent). If the subsets are of equal size, the parent is determined by
+        subset root with the smallest index.
+
+        Parameters
+        ----------
+        a, b : int
+            Node indices to merge.
+
+        Returns
+        -------
+        merged : bool
+            `True` if `a` and `b` were in disjoint sets, `False` otherwise.
+        """
+        if not np.issubdtype(type(a), np.integer):
```

Comment:
I take it you want to limit the API to integers, because it might be useful to restrict the inputs to correspond to indices of an array. Is that right? If so, two immediate questions:

* Right now `d = DisjointSet(); d.find(3)` produces an object whose `n_nodes` is 1; if this were ever changed to be array-based instead, could this be surprising behavior? That is, might one be led to expect that once a query is made for a given input `n`, then `m` is also an available node for all `0 < m < n`?

* Does it make sense that negative inputs are allowed?

Reply:
But at the same time, if the intention is to potentially move to maps whose keys are fixed-size integers, there's potential API breakage here since you currently allow arbitrarily large input values. For complete future-proofing, it seems you would want to restrict them to belong to the appropriate range of values.